File: basic/cs205_c_cpp_2020s/lab11/rusrc/lib/rectangle.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
/*  CS205_C_CPP
    Copyright (C) 2020-2022 nanoseeds
*/
use std::fmt;

pub struct Rectangle {
    pub width: f64,
    pub height: f64,
}

impl Rectangle {
    pub fn get_area(&self) -> f64 {
        self.width * self.height
    }
    pub fn get_perimeter(&self) -> f64 {
        2.0f64 * (self.width + self.height)
    }
    pub fn display(&self) {
        println!("{}", &self);
    }
}

impl fmt::Display for Rectangle {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        return write!(f, r#"    -------------
    Width:      {}
    Height:     {}
    Area:       {:5.3}
    Perimeter:  {:5.3}"#, self.width, self.height, self.get_area(), self.get_perimeter());
    }
}

File: basic/cs205_c_cpp_2020s/lab11/exercise/lib/rectangle.cpp
// SPDX-License-Identifier: AGPL-3.0-or-later
/*  CS205_C_CPP
    Copyright (C) 2020-2023 nanoseeds
*/
/**
 * @Github: https://github.com/Certseeds
 * @Author: nanoseeds
 * @Date: 2020-04-30 21:35:56
 * @LastEditors  : nanoseeds
 */
#include "rectangle.h"
#include <iostream>
#include <iomanip>

rectangle::rectangle() {

}

rectangle::rectangle(double w, double h) : width(w), height(h) {

}

double rectangle::getArea() {
    return this->width * this->height;
}

double rectangle::getPerimeter() {
    return 2 * (this->width + this->height);
}

void rectangle::display() {
    std::cout << "-------------" << std::endl;
    std::cout << "Width:      " << std::setiosflags(std::ios::left) << this->width << std::endl;
    std::cout << "Height:     " << std::setiosflags(std::ios::left) << this->height << std::endl;
    std::cout << "Area:       " << std::setiosflags(std::ios::left) << this->getArea() << std::endl;
    std::cout << "Perimeter:  " << std::setiosflags(std::ios::left) << this->getPerimeter() << std::endl;
}
